README.md
# Snip-It - Code Snippet Generator

Snip-It is a code snippet generator extension created mainly for content creators to create high quality code snippets quickly.

## Features

Snip-it has variety of customization features which lets you customise the snippet according to your need. Some of the features are listed below:


1. Quick Snip

![Quick Snip](https://raw.githubusercontent.com/Rishab49/Rishab49/main/resources/snip-it/first.png)

We can start extension using command palette(`ctrl + shift + p `) and then selecting `start Snipit` from list

2. Change background

![Change background](https://raw.githubusercontent.com/Rishab49/Rishab49/main/resources/snip-it/bg.png)

We can change backgroun color to some common colors to to custom colors using color picker in color tab

3. Change Gradient

![Change gradient](https://raw.githubusercontent.com/Rishab49/Rishab49/main/resources/snip-it/gradient.png)

we can add gradient to the backgrund using gradient tab

4. Change padding

![Change padding](https://raw.githubusercontent.com/Rishab49/Rishab49/main/resources/snip-it/padding.png) 

we ca change padding using padding tab

5. Change label color

![Change label color](https://raw.githubusercontent.com/Rishab49/Rishab49/main/resources/snip-it/text.png)

we can hide or show creator name using text tab and also change its color

6. Image as background

![add image as background ](https://raw.githubusercontent.com/Rishab49/Rishab49/main/resources/snip-it/image.png)

we can add image as background to the snippet using image tab

7. Change theme

![Change theme](https://raw.githubusercontent.com/Rishab49/Rishab49/main/resources/snip-it/theme.png)

we can also change theme using theme drop down.



## How to run

You can start the extension by selecting the code block which you want to convert to snippet and then use `ctrl + shift + p` and then search for `Start Snipit` to start the extension.

## Support & Issues repoting

You can report issue or ask queries on any of my social media account and also you can support my work [here](https://www.buymeacoffee.com/rishabwebdev).

[linkedin](https://www.linkedin.com/in/rishabwebdev/) [twitter](https://twitter.com/RajRishab_) [website](https://rishabraj.me)

## Release Notes

Users appreciate release notes as you update your extension.

### 1.0.0

Initial release of Snip-it
Added features:

1. Quick Snip
2. Change background
3. Change Gradient
4. Change padding
5. Change label color
6. Image as background
7. Change theme

**Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "snip-it" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/test/specs/test.e2e.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const globals_1 = require("@wdio/globals");
const path = require("path");
const webdriverio_1 = require("webdriverio");
describe("VS Code Extension Testing", () => {
    it("should be able to load VSCode", async () => {
        const workbench = await globals_1.browser.getWorkbench();
        (0, globals_1.expect)(await workbench.getTitleBar().getTitle()).toContain("[Extension Development Host]");
    });
});
describe("opening snipit", () => {
    let workbench;
    it("Expect one webview which has title of Snip It", async () => {
        workbench = await globals_1.browser.getWorkbench();
        let dialog = await workbench.executeCommand("workbench.action.files.openFile");
        const input = await dialog.inputBox$.$(dialog.locators.input);
        // await input.click();
        await input.clearValue();
        await input.addValue(path.join(__dirname, "index.html"));
        await dialog.confirm();
        await workbench.getEditorView().openEditor("index.html");
        await globals_1.browser
            .action("key")
            .down(webdriverio_1.Key.ArrowRight)
            .up(webdriverio_1.Key.ArrowRight)
            .perform();
        await workbench.executeCommand("snip-it.start");
        await globals_1.browser.pause(2000);
        await globals_1.browser.waitUntil(async () => {
            let notificaitons = await workbench.getNotifications();
            for (const n of notificaitons) {
                if ((await n.getMessage()).includes("Please select some text")) {
                    n.dismiss();
                    return true;
                }
            }
            return false;
        });
        await workbench.executeCommand("editor.action.selectAll");
        await workbench.executeCommand("snip-it.start");
        await globals_1.browser.waitUntil(async () => (await workbench.getAllWebviews()).length > 0);
        let webviews = await workbench.getAllWebviews();
        webviews[0].open();
    });
    it("frame to have right title", async () => {
        await (0, globals_1.expect)(await (0, globals_1.$)("iframe[title='Snip It']")).toBeTruthy();
    });
    it("svg to have backgrund black", async () => {
        await (0, globals_1.expect)(await (0, globals_1.$)("svg")).toHaveStyle({
            backgroundColor: "rgba(0,0,0,1)",
        });
    });
    it("svg to have 6 childrens", async () => {
        await (0, globals_1.expect)(await (0, globals_1.$)("svg")).toHaveChildren(6);
    });
    it("forignTitle equal to file name", async () => {
        await (0, globals_1.expect)(await (await (0, globals_1.$)("svg")).$(".foreignTitle").$("p")).toHaveText("index.html");
    });
    it("Code element class is right", async () => {
        await (0, globals_1.expect)(await (await (0, globals_1.$)(".foreignCode")).$("code")).toHaveElementClassContaining("language-html");
    });
    it("creator name element has right content and iseditable", async () => {
        await (0, globals_1.expect)(await (await (0, globals_1.$)(".creatorName")).$("p")).toHaveText("@CreatorName");
        await (0, globals_1.expect)(await (await (0, globals_1.$)(".creatorName")).$("p")).toHaveAttr("contenteditable");
    });
    it("aside have all childs", async () => {
        await (0, globals_1.expect)((await (0, globals_1.$)("aside")).$("form")).toHaveChildren(3);
    });
    it("color options have all child", async () => {
        await (0, globals_1.expect)(await (0, globals_1.$)(".color_options")).toHaveChildren(2);
    });
    it("common solid colors are generated", async () => {
        await (0, globals_1.expect)(await (0, globals_1.$)(".common_colors")).toHaveChildren(15);
    });
    it("common gradient are generated", async () => {
        await (0, globals_1.expect)(await (0, globals_1.$)(".common_gradients")).toHaveChildren(10);
    });
    it("gradient-picker to be present", async () => {
        await (0, globals_1.expect)(await (await (0, globals_1.$)(".gradient_colors")).$("gradient-picker")).toBePresent();
    });
    it("color-picker to be present", async () => {
        await (0, globals_1.expect)(await (await (0, globals_1.$)(".solid_colors")).$("color-picker")).toBePresent();
    });
    it("all themes are present", async () => {
        await (0, globals_1.expect)(await (await (0, globals_1.$)(".other_options")).$("select")).toHaveChildren(72);
    });
});
describe("check dynamic features", () => {
    it("should be able to change bg color", async () => {
        let elem = await (0, globals_1.$)(".common_colors").$(".solid_color");
        await elem.click();
        await (0, globals_1.expect)(await (0, globals_1.$)("svg")).toHaveStyle({
            backgroundColor: "rgba(255,0,0,1)",
        });
    });
    // it("should be able to change bg color", async () => {
    //   let elem = await $(".solid_colors").$("color-picker");
    //   let rect = await browser.getElementRect(elem.elementId);
    //   let location = await elem.getLocation();
    //   await elem.click();
    //   await expect(await $("svg")).toHaveStyle({
    //     backgroundColor: "rgba(255,0,0,1)",
    //   });
    //   await browser
    //     .action("key")
    //     .down(Key.ArrowRight)
    //     .pause(5000)
    //     .up(Key.ArrowRight)
    //     .perform();
    //   await browser.pause(5000);
    // });
    it("should be able to change tabs", async () => {
        let elem = await (0, globals_1.$)(".color_options").$("[data-tab='gradient_colors']");
        await elem.click();
        await (0, globals_1.expect)((await (await (0, globals_1.$)(".color_options")).$(".content_container")).$(".gradient_colors")).toHaveElementClassContaining("active");
    });
    it("should be able to change bg gradient", async () => {
        let elem = await (0, globals_1.$)(".common_gradients").$(".gradient_color");
        await elem.click();
        await (0, globals_1.expect)(await (0, globals_1.$)("svg")).toHaveStyle({
            backgroundImage: "linear-gradient(337deg, rgb(101, 78, 163), rgb(218, 152, 180))",
        });
    });
    it("should be able to adjust bg gradient", async () => {
        let elem = await (await (0, globals_1.$)("gradient-picker"))
            .shadow$(".gradient-container")
            .$$(".pin");
        let rect = await globals_1.browser.getElementRect(elem[1].elementId);
        await elem[1].click();
        await (0, globals_1.expect)(elem[1]).toHaveElementClassContaining("active");
        await globals_1.browser
            .action("pointer")
            .move({
            origin: elem[1],
        })
            .down("left")
            .move({
            x: -100,
            y: 0,
            origin: "pointer",
            duration: 5000,
        })
            .up("left")
            .perform();
    });
    it("should be able to change tabs", async () => {
        let elem = await (0, globals_1.$)(".text_options").$("[data-tab='text_style_container']");
        await elem.click();
        await globals_1.browser.pause(10000);
        await (0, globals_1.expect)((await (await (0, globals_1.$)(".text_options")).$(".content_container")).$(".text_style_container")).toHaveElementClassContaining("active");
    });
    it("creator name should be invisible", async () => {
        let label = (await (await (0, globals_1.$)(".text_options")).$(".content_container")).$("label");
        await label.click();
        await (0, globals_1.expect)((0, globals_1.$)(".creatorName").$("p")).toHaveStyle({
            display: "none",
        });
    });
    it("change theme", async () => {
        let select = (await (0, globals_1.$)(".other_options")).$("select");
        select.selectByIndex(1);
        await (0, globals_1.expect)((await (0, globals_1.$)(".foreignCode")).$("code")).toHaveStyle({
            backgroundColor: "rgba(254,254,254,1)",
        });
    });
});
//# sourceMappingURL=test.e2e.js.map

src/test/specs/test.e2e.ts
import { browser, expect, $ } from "@wdio/globals";
import path = require("path");
import fs = require("fs");
import { Workbench } from "wdio-vscode-service";
import { Key } from "webdriverio";
import { describe,it } from "node:test";
describe("VS Code Extension Testing", () => {
  it("should be able to load VSCode", async () => {
    const workbench = await browser.getWorkbench();
    expect(await workbench.getTitleBar().getTitle()).toContain(
      "[Extension Development Host]"
    );
  });
});

describe("opening snipit", () => {
  let workbench: Workbench;
  it("Expect one webview which has title of Snip It", async () => {
    workbench = await browser.getWorkbench();
    let dialog = await workbench.executeCommand(
      "workbench.action.files.openFile"
    );

    const input = await dialog.inputBox$.$(dialog.locators.input);
    // await input.click();
    await input.clearValue();
    await input.addValue(path.join(__dirname, "index.html"));
    await dialog.confirm();

    await workbench.getEditorView().openEditor("index.html");

    await browser
      .action("key")
      .down(Key.ArrowRight)
      .up(Key.ArrowRight)
      .perform();
    await workbench.executeCommand("snip-it.start");

    await browser.pause(2000);

    await browser.waitUntil(async () => {
      let notificaitons = await workbench.getNotifications();
      for (const n of notificaitons) {
        if ((await n.getMessage()).includes("Please select some text")) {
          n.dismiss();
          return true;
        }
      }
      return false;
    });

    await workbench.executeCommand("editor.action.selectAll");
    await workbench.executeCommand("snip-it.start");

    await browser.waitUntil(
      async () => (await workbench.getAllWebviews()).length > 0
    );

    let webviews = await workbench.getAllWebviews();
    webviews[0].open();
  });

  it("frame to have right title", async () => {
    await expect(await $("iframe[title='Snip It']")).toBeTruthy();
  });

  it("svg to have backgrund black", async () => {
    await expect(await $("svg")).toHaveStyle({
      backgroundColor: "rgba(0,0,0,1)",
    });
  });

  it("svg to have 6 childrens", async () => {
    await expect(await $("svg")).toHaveChildren(6);
  });

  it("forignTitle equal to file name", async () => {
    await expect(await (await $("svg")).$(".foreignTitle").$("p")).toHaveText(
      "index.html"
    );
  });

  it("Code element class is right", async () => {
    await expect(
      await (await $(".foreignCode")).$("code")
    ).toHaveElementClassContaining("language-html");
  });

  it("creator name element has right content and iseditable", async () => {
    await expect(await (await $(".creatorName")).$("p")).toHaveText(
      "@CreatorName"
    );
    await expect(await (await $(".creatorName")).$("p")).toHaveAttr(
      "contenteditable"
    );
  });

  it("aside have all childs", async () => {
    await expect((await $("aside")).$("form")).toHaveChildren(3);
  });

  it("color options have all child", async () => {
    await expect(await $(".color_options")).toHaveChildren(2);
  });

  it("common solid colors are generated", async () => {
    await expect(await $(".common_colors")).toHaveChildren(15);
  });

  it("common gradient are generated", async () => {
    await expect(await $(".common_gradients")).toHaveChildren(10);
  });

  it("gradient-picker to be present", async () => {
    await expect(
      await (await $(".gradient_colors")).$("gradient-picker")
    ).toBePresent();
  });

  it("color-picker to be present", async () => {
    await expect(
      await (await $(".solid_colors")).$("color-picker")
    ).toBePresent();
  });

  it("all themes are present", async () => {
    await expect(await (await $(".other_options")).$("select")).toHaveChildren(
      72
    );
  });
});

describe("check dynamic features", () => {
  it("should be able to change bg color", async () => {
    let elem = await $(".common_colors").$(".solid_color");

    await elem.click();

    await expect(await $("svg")).toHaveStyle({
      backgroundColor: "rgba(255,0,0,1)",
    });
  });

  // it("should be able to change bg color", async () => {
  //   let elem = await $(".solid_colors").$("color-picker");
  //   let rect = await browser.getElementRect(elem.elementId);
  //   let location = await elem.getLocation();

  //   await elem.click();

  //   await expect(await $("svg")).toHaveStyle({
  //     backgroundColor: "rgba(255,0,0,1)",
  //   });

  //   await browser
  //     .action("key")
  //     .down(Key.ArrowRight)
  //     .pause(5000)
  //     .up(Key.ArrowRight)
  //     .perform();

  //   await browser.pause(5000);
  // });

  it("should be able to change tabs", async () => {
    let elem = await $(".color_options").$("[data-tab='gradient_colors']");

    await elem.click();

    await expect(
      (
        await (await $(".color_options")).$(".content_container")
      ).$(".gradient_colors")
    ).toHaveElementClassContaining("active");
  });

  it("should be able to change bg gradient", async () => {
    let elem = await $(".common_gradients").$(".gradient_color");

    await elem.click();

    await expect(await $("svg")).toHaveStyle({
      backgroundImage:
        "linear-gradient(337deg, rgb(101, 78, 163), rgb(218, 152, 180))",
    });
  });

  it("should be able to adjust bg gradient", async () => {
    let elem = await (await $("gradient-picker"))
      .shadow$(".gradient-container")
      .$$(".pin");
    let rect = await browser.getElementRect(elem[1].elementId);
    await elem[1].click();

    await expect(elem[1]).toHaveElementClassContaining("active");

    await browser
      .action("pointer")
      .move({
        origin: elem[1],
      })
      .down("left")
      .move({
        x: -100,
        y: 0,
        origin: "pointer",
        duration: 5000,
      })
      .up("left")
      .perform();
  });

  it("should be able to change tabs", async () => {
    let elem = await $(".text_options").$("[data-tab='text_style_container']");

    await elem.click();

    await browser.pause(10000);

    await expect(
      (
        await (await $(".text_options")).$(".content_container")
      ).$(".text_style_container")
    ).toHaveElementClassContaining("active");
  });

  it("creator name should be invisible", async () => {
    let label = (await (await $(".text_options")).$(".content_container")).$(
      "label"
    );
    await label.click();

    await expect($(".creatorName").$("p")).toHaveStyle({
      display: "none",
    });
  });

  it("change theme", async () => {
    let select = (await $(".other_options")).$("select");

    select.selectByIndex(1);

    await expect((await $(".foreignCode")).$("code")).toHaveStyle({
      backgroundColor: "rgba(254,254,254,1)",
    });
  });
});


src/extension.ts
import * as vscode from "vscode";
import { Eta } from "eta";
import { join } from "path";
import { regularTTF, semiTTF, boldTTF } from "./scripts/constant.js";
import {
  createWebviewUri,
  getFileUri,
  readFileContent,
} from "./scripts/methods.js";
export function activate(context: vscode.ExtensionContext) {
  let disposable = vscode.commands.registerCommand(
    "snip-it.start",
    async () => {
      const editor = vscode.window.activeTextEditor;
      let selectedText = editor?.document.getText(editor.selection);
      // console.log(selectedText);
      if(selectedText === null || selectedText?.length === 0){
        vscode.window.showErrorMessage("Please select some text");
        return;
      }
      const panel = vscode.window.createWebviewPanel(
        "SnipIt",
        "Snip It",
        vscode.ViewColumn.One,
        {
          enableScripts: true,
          localResourceRoots: [
            vscode.Uri.file(join(context.extensionPath, "public")),
          ],
        }
      );

      const script = createWebviewUri(context, panel, [
        "public",
        "scripts",
        "highlight.js",
      ]);

      const lineScript = createWebviewUri(context, panel, [
        "public",
        "scripts",
        "linenumber.js",
      ]);

      const customElementScript = createWebviewUri(context, panel, [
        "public",
        "scripts",
        "customElements.js",
      ]);

      const uploadImage = createWebviewUri(context, panel, [
        "public",
        "icons",
        "image.png",
      ]);
      let style = await readFileContent(getFileUri("codepen-embed", context));
      let fileType = editor?.document.fileName.split(".")[1];
      let fileName: string | string[] | undefined = editor?.document.fileName
        .split(".")[0]
        .split("/");
      fileName = fileName?.length ? fileName[fileName?.length - 1] : undefined;

      const eta = new Eta({
        views: join(context.extensionPath, "src", "templates"),
      });

      panel.webview.html = eta.render("./index.html", {
        style: style,
        script: script,
        lineScript: lineScript,
        customElementScript: customElementScript,
        regularTTF: regularTTF,
        semiTTF: semiTTF,
        boldTTF: boldTTF,
        fileType: fileType,
        fileName: fileName + "." + fileType,
        selectedText: "`" + selectedText + "`",
        uploadImage: uploadImage,
      });

      panel.webview.onDidReceiveMessage(async (message) => {
        switch (message.type) {
          case "change_style":
            let style = await readFileContent(
              getFileUri(message.style, context)
            );
            panel.webview.postMessage({
              type: "new_style",
              style: style,
            });
            break;
        }
      });
    }
  );

  context.subscriptions.push(disposable);
}

export function deactivate() {}


webpack.config.js
//@ts-check

'use strict';

const path = require('path');

//@ts-check
/** @typedef {import('webpack').Configuration} WebpackConfig **/

/** @type WebpackConfig */
const extensionConfig = {
  target: 'node', // VS Code extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/
	mode: 'none', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')

  entry: './out/extension.js', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
  output: {
    // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
    path: path.resolve(__dirname, 'dist'),
    filename: 'extension.js',
    libraryTarget: 'commonjs2'
  },
  externals: {
    vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
    // modules added here also need to be added in the .vscodeignore file
  },
  resolve: {
    // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'ts-loader'
          }
        ]
      }
    ]
  },
  devtool: 'nosources-source-map',
  infrastructureLogging: {
    level: "log", // enables logging required for problem matchers
  },
};
module.exports = [ extensionConfig ];

src/templates/script.js
let fileType = "<%= it.fileType %>";
// let bg = "";
let e;
let str = <%= it.selectedText %>;
let padding = {
  left: 32,
  right: 32,
  top: 32,
  bottom: 32,
};
let solidColors = [
  "red",
  "crimson",
  "blue",
  "green",
  "cyan",
  "wheat",
  "whitesmoke",
  "white",
  "gray",
  "darkgray",
  "black",
  "pink",
  "orange",
  "yellow",
  "purple",
];

let linearGradients = [
  "linear-gradient(337deg, rgb(101, 78, 163), rgb(218, 152, 180))",
  "linear-gradient(354deg, rgb(255, 117, 181), rgb(255, 184, 108))",
  "linear-gradient(355deg, rgb(28, 94, 101), rgb(83, 197, 148))",
  "linear-gradient(55deg, rgb(138, 35, 135), rgb(233, 64, 87), rgb(233, 64, 87))",
  "linear-gradient(270deg, rgb(20, 30, 48), rgb(36, 59, 85))",
  "linear-gradient(120deg, rgb(212, 252, 121), rgb(150, 230, 161))",
  "linear-gradient(120deg, rgb(246, 211, 101), rgb(253, 160, 133))",
  "linear-gradient(120deg, rgb(132, 250, 176), rgb(143, 211, 244))",
  "linear-gradient(120deg, rgb(252, 203, 144), rgb(213, 126, 235))",
  "linear-gradient(145deg, rgb(48, 207, 208), rgb(51, 8, 103))",
];



let vscode = acquireVsCodeApi();
let svg = document.querySelector("svg");
let pre = document.querySelector("pre");
let select = document.querySelector("select");
let codeElement = document.querySelector("code");
let bgUpload = document.querySelector(".bg_upload");
let creatorName = document.querySelector(".creatorName");
let foreignCode = document.querySelector(".foreignCode");
let foreignTitle = document.querySelector(".foreignTitle");
let paddingElements = document.querySelectorAll(".padding");
let style_element = document.querySelector("#style_element");
let downloadButton = document.querySelector(".download_button");
let textOptionsContainer = document.querySelector(".text_options");
let colorOptionsContainer = document.querySelector(".color_options");
let linksContainer = colorOptionsContainer.querySelector(".links_container");
let textlinksContainer = textOptionsContainer.querySelector(".links_container");
let contentContainer =
  colorOptionsContainer.querySelector(".content_container");
let textOptionsContentContainer =
  textOptionsContainer.querySelector(".content_container");

let textCheckbox = textOptionsContainer
  .querySelector(".options")
  .querySelector("input[type='checkbox']");
let resizeobserver = new ResizeObserver((e) =>
  e.forEach((entry) => {
    entry.target == pre
      ? setBoundingProps(entry.contentRect.height, entry.contentRect.width)
      : null;
  })
);
downloadButton.addEventListener("click", makeImage);


// window.addEventListener("pointerdown",(event) => {
//   e = document.createElement("div");
//   e.style.position = "absolute";
//   e.style.height="50px";
//   e.style.width="50px";
//   e.style.zIndex="10000";
//   e.style.backgroundColor="red";
//   e.style.transform=`translate(${event.x}px,${event.y}px)`;
  
//   document.body.appendChild(e);
// });

// window.addEventListener("pointermove",(event) => {
//   e.style.transform = `translate(${event.x}px,${event.y}px)`;
// })
// window.addEventListener("pointerup",(event) => {
//   document.body.removeChild(e);
// })

window.addEventListener("load", () => {
  resizeobserver.observe(pre);
  codeElement.innerHTML = str;
  hljs.highlightAll();
  hljs.initLineNumbersOnLoad();
  createSolidColorElements();
  createGradientColorElements();
});

select.addEventListener("change", (event) =>
  vscode.postMessage({
    type: "change_style",
    style: event.target.value,
  })
);

bgUpload.addEventListener("change", (event) => getBase64(bgUpload.files[0]));

linksContainer.querySelectorAll("li").forEach((li) => {
  li.addEventListener("click", (e) => {
    contentContainer
      .querySelectorAll("div")
      .forEach((div) =>
        div.classList.contains(e.target.getAttribute("data-tab"))
          ? div.classList.add("active")
          : div.classList.remove("active")
      );
    linksContainer
      .querySelectorAll("li")
      .forEach((li) =>
        li == e.target
          ? li.classList.add("active")
          : li.classList.remove("active")
      );
  });
});

textlinksContainer.querySelectorAll("li").forEach((li) => {
  li.addEventListener("click", (e) => {
    textOptionsContentContainer.querySelectorAll("div").forEach((div) => {
      div.classList.contains(e.target.getAttribute("data-tab"))
        ? div.classList.add("active")
        : div.classList.remove("active");
    });
    textlinksContainer.querySelectorAll("li").forEach((li) => {
      li == e.target
        ? li.classList.add("active")
        : li.classList.remove("active");
    });
  });
});

textCheckbox.addEventListener("change", (event) => {
  event.target.checked
    ? (creatorName.querySelector("p").style.display = "block")
    : (creatorName.querySelector("p").style.display = "none");
});

paddingElements.forEach((elem) => {
  elem.querySelector("p").addEventListener("input", (event) => {
    let className = elem.getAttribute("class").split(" ")[1];
    padding[className] = Number(event.target.innerText);
    setBoundingProps(
      pre.getBoundingClientRect().height,
      pre.getBoundingClientRect().width
    );
    if (className == "bottom") {
      creatorName.setAttribute("height", padding["bottom"] - 10);
    }
  });
});

window.addEventListener("message", (event) => {
  let message = event.data;
  switch (message.type) {
    case "new_style":
      svg.querySelector("style")
        ? svg.removeChild(svg.querySelector("style"))
        : null;
      style_element.innerHTML = message.style;
      break;
  }
});

async function makeImage() {
  let canvas = document.querySelector("canvas");

  const style = createStyleElementFromCSS();
  svg.insertBefore(style, svg.firstChild);

  let xml = new XMLSerializer().serializeToString(svg);
  let svg64 = btoa(xml);
  let b64start = "data:image/svg+xml;base64,";

  let img64 = b64start + svg64;
  let img = new Image();

  let bgImg = new Image();
  bgImg.src = bg;
  canvas.height = svg.clientHeight * 3;
  canvas.width = svg.clientWidth * 3;

  let ratio = canvas.width / canvas.height;
  img.onload = function () {
    let ctx = canvas.getContext("2d");
    ctx.drawImage(
      bgImg,
      0,
      0,
      bgImg.width,
      bgImg.height,
      0,
      0,
      canvas.height * ratio,
      canvas.height
    );
    ctx.drawImage(
      img,
      0,
      0,
      img.width,
      img.height,
      0,
      0,
      canvas.height * ratio,
      canvas.height
    );

    downloadImg();
  };

  img.src = img64;
}

const createStyleElementFromCSS = () => {
  let elemObj = {
    name: "style",
    type: "text/css",
  };

  const style = createElement(elemObj);
  const sheet = [...document.styleSheets];

  sheet.forEach((s) => {
    const styleRules = [];
    for (let i = 0; i < s.cssRules.length; i++)
      styleRules.push(s.cssRules.item(i).cssText);
    style.appendChild(document.createTextNode(styleRules.join(" ")));
  });
  return style;
};

function downloadImg() {
  let canvas = document.querySelector(" canvas");
  let data = canvas.toDataURL();

  let elemObj = {
    name: "a",
    href: data,
    attributes: {
      download: "snipit.jpg",
    },
  };

  let elem = createElement(elemObj);
  elem.href = data;
  elem.setAttribute("download", "snipit.jpg");
  elem.click();
}
function getBase64(file) {
  let reader = new FileReader();
  reader.readAsDataURL(file);
  reader.onload = function () {
    bg = reader.result;
    svg.style.backgroundImage = "url(" + bg + ")";
    svg.style.backgroundPosition = "center";
    svg.style.backgroundSize = "cover";
  };
}
function createSolidColorElements() {
  let solidColorsElement = document
    .querySelector(".solid_colors")
    .querySelector(".common_colors");
  solidColors.forEach((color) => {
    let elemObj = {
      name: "div",
      attributes: {
        class: "solid_color",
      },
      styles: {
        backgroundColor: color,
      },
      eventListeners: {
        click: () => (svg.style.background = color),
      },
    };

    let elem = createElement(elemObj);
    solidColorsElement.appendChild(elem);
  });
}

function createGradientColorElements() {
  let gradientColorsElement = document
    .querySelector(".gradient_colors")
    .querySelector(".common_gradients");
  linearGradients.forEach((color) => {
    let elemObj = {
      name: "div",
      attributes: {
        class: "gradient_color",
      },
      styles: {
        backgroundImage: color,
      },
      eventListeners: {
        click: () => (svg.style.backgroundImage = color),
      },
    };
    let elem = createElement(elemObj);
    gradientColorsElement.appendChild(elem);
  });
}

function setBoundingProps(height, width) {
  foreignCode.setAttribute("height", height);
  foreignCode.setAttribute("width", width);

  foreignTitle.setAttribute("width", width);

  creatorName.setAttribute("y", height + 35);
  creatorName.setAttribute("width", width);

  svg.setAttribute("height", height + 35 + padding.top + padding.bottom);
  svg.setAttribute("width", width + padding.left + padding.right);
  svg.setAttribute(
    "viewBox",
    `-${padding.left} -${padding.top} ` +
      (width + padding.left + padding.right) +
      " " +
      (height + padding.top + padding.bottom + 35)
  );
}

function createElement(options) {
  let elem = document.createElement(options.name);
  for (let option in options.attributes) {
    elem.setAttribute(option, options.attributes[option]);
  }

  for (let option in options.styles) {
    elem.style[option] = options.styles[option];
  }

  for (let option in options.eventListeners) {
    elem.addEventListener(option, options.eventListeners[option]);
  }

  options.type ? (elem.type = options.type) : null;
  options.href ? (elem.href = options.href) : null;
  return elem;
}
